*** vaxuba/ubavar.h.4.3	Thu Jun  5 04:20:06 1986
--- vaxuba/ubavar.h	Tue Feb 10 20:56:55 1987
***************
*** 60,64 ****
  };
  
- #ifndef LOCORE
  /*
   * Per-controller structure.
--- 60,63 ----
***************
*** 83,86 ****
--- 82,86 ----
  	int	um_cmd;		/* communication to dgo() */
  	int	um_ubinfo;	/* save unibus registers, etc */
+ 	int	um_bdp;		/* for controllers that hang on to bdp's */
  	struct	buf um_tab;	/* queue of devices for this controller */
  };
***************
*** 121,125 ****
  	struct	uba_hd *ui_hd;
  };
- #endif
  
  /*
--- 121,124 ----
***************
*** 141,144 ****
--- 140,144 ----
  	struct	uba_ctlr **ud_minfo;	/* backpointers to ubminit structs */
  	short	ud_xclu;		/* want exclusive use of bdp's */
+ 	short	ud_keepbdp;		/* hang on to bdp's once allocated */
  	int	(*ud_ubamem)();		/* see if dedicated memory is present */
  };
***************
*** 164,167 ****
--- 164,169 ----
  #ifndef LOCORE
  #ifdef KERNEL
+ #define	ubago(ui)	ubaqueue(ui, 0)
+ 
  /*
   * UBA related kernel variables
