*** if.h.orig	Mon Mar 24 15:24:20 1986
--- if.h	Mon Mar 24 15:24:22 1986
***************
*** 75,80 ****
--- 75,81 ----
  #define	IFF_NOTRAILERS	0x20		/* avoid use of trailers */
  #define	IFF_RUNNING	0x40		/* resources allocated */
  #define	IFF_NOARP	0x80		/* no address resolution protocol */
+ #define	IFF_SUBARP	0x200		/* ARP subnet gateway hack */
  /* flags set internally only: */
  #define	IFF_CANTCHANGE	(IFF_BROADCAST | IFF_POINTOPOINT | IFF_RUNNING)
  
