*** in.h.orig	Sun Apr  6 14:43:09 1986
--- in.h	Sun Apr  6 14:43:11 1986
***************
*** 104,108 ****
  extern	struct domain inetdomain;
  extern	struct protosw inetsw[];
  struct	in_addr in_makeaddr();
! u_long	in_netof(), in_lnaof();
  #endif
--- 104,108 ----
  extern	struct domain inetdomain;
  extern	struct protosw inetsw[];
  struct	in_addr in_makeaddr();
! u_long	in_netof(), in_lnaof(), in_mainnetof();
  #endif
