*** vaxif/if_ec.c.4.3	Thu Jun  5 20:04:36 1986
--- vaxif/if_ec.c	Tue Feb 10 20:57:32 1987
***************
*** 58,62 ****
  u_short ecstd[] = { 0 };
  struct	uba_driver ecdriver =
! 	{ ecprobe, 0, ecattach, 0, ecstd, "ec", ecinfo, 0, 0, 0, ecubamem };
  
  int	ecinit(),ecioctl(),ecoutput(),ecreset();
--- 58,62 ----
  u_short ecstd[] = { 0 };
  struct	uba_driver ecdriver =
! 	{ ecprobe, 0, ecattach, 0, ecstd, "ec", ecinfo, 0, 0, 0, 0, ecubamem };
  
  int	ecinit(),ecioctl(),ecoutput(),ecreset();
